packages/react-duckdb/src/duckdb_provider.tsx

```diff
@@ -0,0 +1,184 @@
+"use client";
+
+import React, { createContext, useCallback, useContext, useEffect, useState } from 'react';
+import {
+    AsyncDuckDB,
+    type AsyncDuckDBConnection,
+    ConsoleLogger,
+    type DuckDBBundle,
+    type DuckDBBundles,
+    type DuckDBConfig,
+    type InstantiationProgress,
+    type Logger,
+    selectBundle,
+} from '@duckdb/duckdb-wasm';
+import { Resolvable } from './resolvable';
+
+function isDuckDBBundle(bundle: unknown): bundle is DuckDBBundle {
+    return (
+        typeof bundle === 'object' &&
+        bundle !== null &&
+        'mainModule' in bundle &&
+        typeof bundle.mainModule === 'string' &&
+        'mainWorker' in bundle &&
+        (typeof bundle.mainWorker === 'string' || bundle.mainWorker === null) &&
+        'pthreadWorker' in bundle &&
+        (typeof bundle.pthreadWorker === 'string' || bundle.pthreadWorker === null)
+    );
+}
+
+type ConnectionPool = Record<number, AsyncDuckDBConnection>;
+
+type DuckDBContextValue = {
+    database: AsyncDuckDB | null;
+    connectionPool: ConnectionPool;
+    establishConnection: (id?: number) => Promise<void>;
+};
+
+const DuckDBContext = createContext<DuckDBContextValue | null>(null);
+
+// TODO: consider adding support for passing in an existing AsyncDuckDB instance
```

Comment:
My $0.02 is it seems like most of the complexity is to handle the `bundle`, `bundles`, etc

I think this is a wise comment, and I even wonder if supporting _only_ an existing `AsyncDuckDB` instance makes sense